### PR TITLE
Use rust-bitcoin-maintainer-tools and re-write CI

### PR DESCRIPTION
**Pull request by tcharding**
As we have been doing in other crates, use the new maintainer tools test script we created from `rust-bitcoin/rust-bitcoin-maintainer-tools/ci`.

- This should not change test coverage in any meaningful way.
- The integration tests are not currently run in CI, this patch preserves that behaviour.
- Introduces recent/minimal lock files.

### CODE REVIEW

**Comment by apoelstra** [comment]
Looks like the `serde_json` version is too low and some other crates' versions are too high, in the lockfiles.

We also probably want to pin nightly rustc in the same PR.

**Comment by tcharding** [comment]
Note please the recent and minimal lock files are the same. I spent 10 minutes trying to fix them otherwise, but my fingers started to bleed.

Also, I'll fix the lint job after #117 goes in.

**Comment by tcharding** [comment]
I don't see whats up with the lint job, I can't reproduce locally?

**Comment by apoelstra** [comment]
@tcharding it looks like your lockfiles have `serde` 1.0.101, which does not compile anymore.

It looks like serde 1.0.103 will work, which is what we have in the Cargo.toml for `bitcoin` and `units` and `internals` (`hashes` has 1.0 which is a bug, oops). I remember the minimal version needing to be even higher...our Cargo-minimal.lock in the rust-bitcoin repo is at 1.0.156, which is suspicious because the minimal lockfile should match Cargo.toml.

Anyway, to get this PR compiling we just need to bump the serde version. But separately I should go audit the lockfiles in rust-secp and rust-bitcoin.

**Comment by tcharding** [comment]
Since this crate is MSRV 1.48 still, I wasn't able to use the other repos lock files for guidance. Simply using `cargo update -p serde --precise 1.0.103` doesn't seem to work here. I'm on an epic fast to try get over all the illnesses I keep getting clobbered with at the moment so I'll come back to this when I'm eating again.

**Comment by tcharding** [comment]
Oh the MSRV thing is going to be a problem if we want to move this to `bitcoind-json-rpc` repo, any reason not to bump this crate to 1.56.1?

**Comment by apoelstra** [comment]
Nope, no reason. Let's do it.

**Comment by tcharding** [comment]
Now includes the MSRV bump, are you ok to do that in this PR or is that too much going on?

**Comment by tcharding** [comment]
Not sure whats up with the lint fail.

**Comment by apoelstra** [comment]
The lint failure looks like an outdated version of serde is being used.

**Comment by tcharding** [comment]
oh thanks, nice. I'll try to fix that up.

**Comment by apoelstra** [comment]
Still having trouble with your lockfiles. I will try to figure it out and submit a diff.

File: integration_test/Cargo.toml [hunk 1]
```diff
@@ -8,5 +8,5 @@ edition = "2018"
 jsonrpc = { path = "..", features = ["minreq_http"] }
 lazy_static = "1.4.0"
 log = "0.4"
-backtrace = "0.3.50"
+backtrace = "0.2.3"
```
Comment (apoelstra) [inline integration_test/Cargo.toml:11]: In 10972a7846412d323911b382010a657a539b5296

Why did you change this? backtrace 0.2 does not link for me.

Comment (tcharding) [reply integration_test/Cargo.toml]: No idea what I was doing, now removed. Note the recent and minimal files are the same.

**Comment by apoelstra** [comment]
OK, this is looking better. The new lockfiles still don't work with 1.48.0 but at least this is consistent when running locally `cargo +1.48.0 test` and in CI, rather than being a mysterious nix-only thing.

You will probably have a better time with this if you first bump the MSRV to 1.56.1.

**Comment by tcharding** [comment]
Oh I did not realize we were on 1.48 here still. I'll bump to 1.63

**Comment by tcharding** [comment]
Rebased to pick up the MSRV bump. Also added a patch upgrading `honggfuzz`.

File: integration_test/Cargo.toml (hunk 1 above)
Comment (apoelstra) [reply integration_test/Cargo.toml:11]: In a065c9fe4b009dc2079cbb2970a83213b602b1cb:

This is not removed.

File: contrib/crates.sh [hunk 2]
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env bash
+
+# Crates in this workspace to test (excl. fuzz an integration-tests).
+CRATES=(".")                    # Non-workspaces don't have crates.
```
Comment (apoelstra) [inline contrib/crates.sh:4]: In a065c9fe4b009dc2079cbb2970a83213b602b1cb:

What is going on with this file? What does "non-workspaces don't have crates" mean, and does this actually exclude testing the `fuzz` and `integration-test` crates, and why?

Comment (tcharding) [reply contrib/crates.sh:4]: Oh yes you are correct this should include `fuzz`, and I have that wrong in `miniscript` as well. The comment is also wrong and was always wrong, I wrote "Non-workspcase" when I meant "repos that have crate root in the same directory as repo root" (I don't know the name for that). We need to keep the "." so it should be 
```
CRATES=("." "fuzz")     # Some comment about the dot directory.
```

Comment (apoelstra) [reply contrib/crates.sh:4]: I think you want then

```
# Directories of each workspace to test
CRATES=("." "fuzz")
```

Saying "crates in this workspace" is just confusing. Workspaces are crates.

Comment (apoelstra) [reply contrib/crates.sh:4]: BTW the reason I'm complaining about this is that the existing lockfiles don't work with MSRV, which you can tell if you try to compile the `integration_test` workspace.

Comment (tcharding) [reply contrib/crates.sh:4]: I think I also forgot that this crate is 1.48 still, at least that was not in my mind yesterday.

Comment (tcharding) [reply contrib/crates.sh:4]: oh there's the problem, I bumped the MSRV in this PR - I'll pull it out into a separate one.

Comment (tcharding) [reply contrib/crates.sh:4]: Alright, I'm well confused. I copied `Cargo-recent.lock` to `Cargo.lock` and cd'ed into `integration_test` and ran
```
 cargo +1.63.0 --locked check
```

And it was ok.

Comment (tcharding) [reply contrib/crates.sh]: I just went right ahead and added `fuzz` and `integration_test`  to `CRATES` also.

Note that we do not run the integration tests now or previously.

Comment (tcharding) [reply contrib/crates.sh]: > I think I also forgot that this crate is 1.48 still

What the hell was I thinking

File: integration_test/Cargo.toml (hunk 1 above)
Comment (apoelstra) [reply integration_test/Cargo.toml:11]: In d4ed55570290e378aee957fdf006d0ce0bb1749b:

Please remove this.

Comment (tcharding) [reply integration_test/Cargo.toml]: My bad. Done, again.

**Comment by apoelstra** [comment]
If you run `cargo +1.63.0 test --all-features --locked` in the integration test directory, none of `object` `backtrace` `addr2line` (at least) compile. In 04dc4b2ffe6af33feb5481e65bc03f3b9b09ae9a.

**Comment by tcharding** [comment]
hmm, something is not quite right then because that should have been caught by CI. I'll investigate further.

**Comment by tcharding** [comment]
Found the bug in `run_task`, we have a `break` instead of a `continue` in a loop. Funny after we just hit that same mistake in the bech32 the other day. We were not hitting it before because `fuzz` was always at the end of the list of `CRATES`.

**Comment by apoelstra** [comment]
Oh, neat. I feel much better about what a PITA this PR has been. Apparently we needed to use the maintainer tools on a crate whose CI was a total mess :).

**Comment by tcharding** [comment]
Ok, this one should be right to merge if it gets past your local CI.

**Review by apoelstra — APPROVED** [review]
ACK 8f1779447131de57e0f32797bfa16209694d1208 successfully ran local tests
